doc/show_comics.1
.rn '' }`
''' $RCSfile$$Revision$$Date$
'''
''' $Log$
''' Revision 1.1  2000-04-26 19:07:12  hochstrb
''' Initial revision
'''
'''
.de Sh
.br
.if t .Sp
.ne 5
.PP
\fB\\$1\fR
.PP
..
.de Sp
.if t .sp .5v
.if n .sp
..
.de Ip
.br
.ie \\n(.$>=3 .ne \\$3
.el .ne 3
.IP "\\$1" \\$2
..
.de Vb
.ft CW
.nf
.ne \\$1
..
.de Ve
.ft R

.fi
..
'''
'''
'''     Set up \*(-- to give an unbreakable dash;
'''     string Tr holds user defined translation string.
'''     Bell System Logo is used as a dummy character.
'''
.tr \(*W-|\(bv\*(Tr
.ie n \{\
.ds -- \(*W-
.ds PI pi
.if (\n(.H=4u)&(1m=24u) .ds -- \(*W\h'-12u'\(*W\h'-12u'-\" diablo 10 pitch
.if (\n(.H=4u)&(1m=20u) .ds -- \(*W\h'-12u'\(*W\h'-8u'-\" diablo 12 pitch
.ds L" ""
.ds R" ""
.ds L' '
.ds R' '
'br\}
.el\{\
.ds -- \(em\|
.tr \*(Tr
.ds L" ``
.ds R" ''
.ds L' `
.ds R' '
.ds PI \(*p
'br\}
.\"	If the F register is turned on, we'll generate
.\"	index entries out stderr for the following things:
.\"		TH	Title 
.\"		SH	Header
.\"		Sh	Subsection 
.\"		Ip	Item
.\"		X<>	Xref  (embedded
.\"	Of course, you have to process the output yourself
.\"	in some meaninful fashion.
.if \nF \{
.de IX
.tm Index:\\$1\t\\n%\t"\\$2"
..
.nr % 0
.rr F
.\}
.TH SHOW_COMICS 1 "netcomics-0.12" "26/Apr/100" "Web Utilities"
.IX Title "SHOW_COMICS 1"
.UC
.IX Name "show_comics - Show comics, distributed over a time interval"
.if n .hy 0
.if n .na
.ds C+ C\v'-.1v'\h'-1p'\s-2+\h'-1p'+\s0\v'.1v'\h'-1p'
.de CQ          \" put $1 in typewriter font
.ft CW
'if n "\c
'if t \\&\\$1\c
'if n \\&\\$1\c
'if n \&"
\\&\\$2 \\$3 \\$4 \\$5 \\$6 \\$7
'.ft R
..
.\" @(#)ms.acc 1.5 88/02/08 SMI; from UCB 4.2
.	\" AM - accent mark definitions
.bd B 3
.	\" fudge factors for nroff and troff
.if n \{\
.	ds #H 0
.	ds #V .8m
.	ds #F .3m
.	ds #[ \f1
.	ds #] \fP
.\}
.if t \{\
.	ds #H ((1u-(\\\\n(.fu%2u))*.13m)
.	ds #V .6m
.	ds #F 0
.	ds #[ \&
.	ds #] \&
.\}
.	\" simple accents for nroff and troff
.if n \{\
.	ds ' \&
.	ds ` \&
.	ds ^ \&
.	ds , \&
.	ds ~ ~
.	ds ? ?
.	ds ! !
.	ds /
.	ds q
.\}
.if t \{\
.	ds ' \\k:\h'-(\\n(.wu*8/10-\*(#H)'\'\h"|\\n:u"
.	ds ` \\k:\h'-(\\n(.wu*8/10-\*(#H)'\`\h'|\\n:u'
.	ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'^\h'|\\n:u'
.	ds , \\k:\h'-(\\n(.wu*8/10)',\h'|\\n:u'
.	ds ~ \\k:\h'-(\\n(.wu-\*(#H-.1m)'~\h'|\\n:u'
.	ds ? \s-2c\h'-\w'c'u*7/10'\u\h'\*(#H'\zi\d\s+2\h'\w'c'u*8/10'
.	ds ! \s-2\(or\s+2\h'-\w'\(or'u'\v'-.8m'.\v'.8m'
.	ds / \\k:\h'-(\\n(.wu*8/10-\*(#H)'\z\(sl\h'|\\n:u'
.	ds q o\h'-\w'o'u*8/10'\s-4\v'.4m'\z\(*i\v'-.4m'\s+4\h'\w'o'u*8/10'
.\}
.	\" troff and (daisy-wheel) nroff accents
.ds : \\k:\h'-(\\n(.wu*8/10-\*(#H+.1m+\*(#F)'\v'-\*(#V'\z.\h'.2m+\*(#F'.\h'|\\n:u'\v'\*(#V'
.ds 8 \h'\*(#H'\(*b\h'-\*(#H'
.ds v \\k:\h'-(\\n(.wu*9/10-\*(#H)'\v'-\*(#V'\*(#[\s-4v\s0\v'\*(#V'\h'|\\n:u'\*(#]
.ds _ \\k:\h'-(\\n(.wu*9/10-\*(#H+(\*(#F*2/3))'\v'-.4m'\z\(hy\v'.4m'\h'|\\n:u'
.ds . \\k:\h'-(\\n(.wu*8/10)'\v'\*(#V*4/10'\z.\v'-\*(#V*4/10'\h'|\\n:u'
.ds 3 \*(#[\v'.2m'\s-2\&3\s0\v'-.2m'\*(#]
.ds o \\k:\h'-(\\n(.wu+\w'\(de'u-\*(#H)/2u'\v'-.3n'\*(#[\z\(de\v'.3n'\h'|\\n:u'\*(#]
.ds d- \h'\*(#H'\(pd\h'-\w'~'u'\v'-.25m'\f2\(hy\fP\v'.25m'\h'-\*(#H'
.ds D- D\\k:\h'-\w'D'u'\v'-.11m'\z\(hy\v'.11m'\h'|\\n:u'
.ds th \*(#[\v'.3m'\s+1I\s-1\v'-.3m'\h'-(\w'I'u*2/3)'\s-1o\s+1\*(#]
.ds Th \*(#[\s+2I\s-2\h'-\w'I'u*3/5'\v'-.3m'o\v'.3m'\*(#]
.ds ae a\h'-(\w'a'u*4/10)'e
.ds Ae A\h'-(\w'A'u*4/10)'E
.ds oe o\h'-(\w'o'u*4/10)'e
.ds Oe O\h'-(\w'O'u*4/10)'E
.	\" corrections for vroff
.if v .ds ~ \\k:\h'-(\\n(.wu*9/10-\*(#H)'\s-2\u~\d\s+2\h'|\\n:u'
.if v .ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'\v'-.4m'^\v'.4m'\h'|\\n:u'
.	\" for low resolution devices (crt and lpr)
.if \n(.H>23 .if \n(.V>19 \
\{\
.	ds : e
.	ds 8 ss
.	ds v \h'-1'\o'\(aa\(ga'
.	ds _ \h'-1'^
.	ds . \h'-1'.
.	ds 3 3
.	ds o a
.	ds d- d\h'-1'\(ga
.	ds D- D\h'-1'\(hy
.	ds th \o'bp'
.	ds Th \o'LP'
.	ds ae ae
.	ds Ae AE
.	ds oe oe
.	ds Oe OE
.\}
.rm #[ #] #H #V #F C
.SH "NAME"
.IX Header "NAME"
show_comics \- Show comics, distributed over a time interval
.SH "SYNOPSIS"
.IX Header "SYNOPSIS"
\fBshow_netcomics\fR [\fB\-c\fR \fIcmd\fR] [\fB\-d\fR \fIdir\fR] [\fB\-f\fR \fIrcfile\fR] [\fB\-l\fR|\fB\-t\fR \fIhh:mm\fR] [\fB\-p\fR [cmd]] [\fB\-k\fR \fIlockfile\fR|\fB\-K\fR] [\fB\-s\fR <hh:mm>]
.SH "DESCRIPTION"
.IX Header "DESCRIPTION"
show_comics is used to display comics downloaded by netcomics.  It
first pops up a window to ask you what time you plan on leaving (in 24
hour time).  Using the time you entered, it will evenly distribute the
comics to be displayed from the time the script is run to the time you
specified you will leave.
.PP
Specify the comics you want to have displayed in a file called
~/.comics.  You may specify an alternate file on the command line.  If no
file exists, all comics will be displayed (all files that end 
with .gif, .jpg, or .jpeg in the directory).
.PP
If you don't have perlTk installed, use \fB\-l\fR to specify the time and to make
show_comics not try to pop up a window.
.PP
In a UNIX environment, you can utilize the lockfile mechanism to run
show_comics from many different places an not have to worry about
checking to make sure it is already running. For example, you can run
show_comics from your .xinitrc or .xsession as well as from a
crontab. If netcomics is run in the morning, say at 7:00 by a crontab,
then you could run show_comics from a cron job at 7:30, assuming
netcomics will take no more than 1/2 hour. Then, in your .xinitrc
file, you could also run show_comics if the time is greater than 7:30
(see \fB\-s\fR). Therefore, if you have logged into your computer before
7:30, show_comics won't run from your .xinitrc file, but once 7:30 has
come around, it will be run by your cron job. If you log into your
computer after 7:30, then show_comics will fail to run at 7:30
(assuming the program show_comics uses to check to see if the display
is available failed), and then show_comics will be run by your
.xinitrc file. 
.SH "OPTIONS"
.IX Header "OPTIONS"
.Ip "\fB\-c\fR \fIcommand\fR" 4
.IX Item "\fB\-c\fR \fIcommand\fR"
Use the specified command to display each comic instead of the default: 
\*(L"nice \-19 xv\*(R".
.Ip "\fB\-d\fR \fIdir\fR" 4
.IX Item "\fB\-d\fR \fIdir\fR"
Use the specified location of the comics to be displayed instead of using
the default directory: /home/benh/www/comics.
.Ip "\fB\-f\fR \fIfilename\fR" 4
.IX Item "\fB\-f\fR \fIfilename\fR"
Use the specified file instead of ~/.comics for the list of comics to
be displayed
.Ip "\fB\-k\fR \fIfilename\fR" 4
.IX Item "\fB\-k\fR \fIfilename\fR"
Specify the lock file's name.  Default is ~/.show_comics_lock.
.Ip "\fB\-l\fR \fIhh:mm\fR" 4
.IX Item "\fB\-l\fR \fIhh:mm\fR"
Don't pop up a window and use the specified time as the time to leave.
.Ip "\fB\-p\fR [\fIcmd\fR]" 4
.IX Item "\fB\-p\fR [\fIcmd\fR]"
Check to make sure the display is available and if not, quit. Optional
argument is the name of the program to use.  A non-zero return code
from this program indicates a failure to open the display. The default
program used is xdpyinfo.
.Ip "\fB\-s\fR \fIhh:mm\fR" 4
.IX Item "\fB\-s\fR \fIhh:mm\fR"
Specify a time that if show_comics is run before during the day, it will just 
exit.
.Ip "\fB\-t\fR \fIhh:mm\fR" 4
.IX Item "\fB\-t\fR \fIhh:mm\fR"
Specify the default time to be put in the pop-up window.
.Ip "\fB\-v\fR   Output verbose messages." 4
.IX Item "\fB\-v\fR   Output verbose messages."
.SH "FILES"
.IX Header "FILES"
.Ip "~/.comics" 33
.IX Item "~/.comics"
Contains the a list of comics you want displayed, one per line.
.Ip "/home/benh/www/comics" 33
.IX Item "/home/benh/www/comics"
Default directory where comics and the web page are placed by netcomics.
.Ip "~/.show_comics_lock" 33
.IX Item "~/.show_comics_lock"
Default lock file used to make sure only one instance of show_comics is run
.SH "SEE ALSO"
.IX Header "SEE ALSO"
\fInetcomics\fR\|(1)
.SH "AUTHOR"
.IX Header "AUTHOR"
Ben Hochstedler <hochstrb@cs.rose-hulman.edu>
ICQ: \fB15469308\fR AIM: \fBhochstrb\fR

.rn }` ''


NEWS
Changes in release 0.13.1:
* Several bug fixes.
* The option -a ($always_download) when used with -c now only forces
  download of the specified comics.  Failed comics that haven't
  reached the maximum retry limit are also retried.
* New comics in this release:
  bignate     Big Nate
  bigp        The Big Picture
  bz          Bizarro
  bobo        Bobos Progress
  bd          The Boondocks
  bullsnbears Bulls n Bears 
  cwbw        Cartoon Web: BizWit
  cwc         Cartoon Web: Caricatures
  cwet        Cartoon Web: Ecotoon
  kal         Cartoon Web: Kal
  cws         Cartoon Web: Symbolics
  votw        Cartoon Web: Views of the World
  wotw        Cartoon Web: Wit of the World
  cwww        Cartoon Web: WitWomen
  wvotd       Cartoon Web: World View of the day
  cheap       Cheap Thrills Cuisine
  dork        Dork Tower
  exploit     Exploitation Now
  fatjc       Fat Jesus
  fce         Fight Cast Or Evade
  flightdeck  Flight Deck
  grandave    Grand Avenue
  grandclause Grandfather Clause
  jm          James
  lupo        Lupo Alberto
  mrlowe      Mr. Lowe
  pearls      Pearls Before Swine
  pickles     Pickles
  duncan      Raising Duncan
  randolph    Randolph Itch 2:00 AM
  reality     Reality Check
  redrover    Red and Rover
  shirley     Shirley and Son
  soupnutz    Soup to Nutz
  srcwars     Source Wars
* Several comics have been removed either because the artist requested
  it, or the site from which it was being retreaved dropped it.

Major changes in release 0.13:

* Each comics strip's Resource Locator Info (RLI) structure is now
  saved persistently in the spool directory when the comic strip is
  downloaded.  These "rli status files" are reloaded during the next
  invocation of netcomics (unless they're purged) and used to
  determine what comics needs to be retried to be downloaded.  When
  the webpage is recreated, the status files are used to find out
  extra information about the already downloaded comics (author,
  caption, website URLs, etc).  New commandline options: -R -a -q.
* Referers are now handled.  A module may specify its own referer if
  the default (derived from the URL) is not correct.
* The external command specified may now include several macros which
  get replaced.  See the manpage for details.
* New comics in this release: 
  btaf		Bob the Angry Flower
  cotc		Clan of the Cats
  crfh		College Roomies from Hell
  drfun		Dr. Fun
  evjake	Everything Jake
  elf		Elf Life
  gu		Gaming U
  bayside	Life at Bayside
  nukees	Nukees
  sempai	Sempai
  sinf		Sinfest
  wigu		Whenigrowup

Major changes in release 0.12:

* Added the comics: 9 Chickweed Lane, Absurd Notion, Against The
  Grain, Badlands, Bobos Progress, Boonies, Broom-Hilda, Buckets, Cats
  With Hands, Chase Villens, Dilbert Classics, Down To Earth, Dr Katz,
  Eek and Meek, Fair Game, Gibberish, Gil Thorp, Ick, Janes World, The
  Japanese Beetle, K Rat, Lola, Loose Parts, Mary Worth, Meatloaf
  Night, Mickey Mouse, The New Breed, Outtake, PC and Pixel, PS
  Mueller, Penny Arcade, Penny Arcade Bench, Phantom, Player Versus
  Player, Pretzel Logic, Quigmans, Random Shots, Rex Morgan, Ripleys
  Believe It Or Not, Rubes, Small World, Spex and Wally, Staggering
  Heights, Suburban Jungle, Sylvia, Tarzan, Thats Life, Top Of This
  World, Toy Trunk Railroad, WildLife, You Damn Kid, Kudzu, Small
  Grey, and Fat Cats.
* Fixed Y2K bugs.

Major changes in release 0.11:

* Added the comics: After Y2K, Dysentery, Cartoon Web, Bound & Gagged,
  Fred Basset, Herb & Jaamal, Middletons, Pluggers, Motley's Crew,
  Zorro, Liberty Meadows, Teenage Ninja Turtles, Ballard Street, Miss
  Peach, Wee Pals, Raw Material, Agnes, Dr Katz, Outtake, Pcpixel, 9
  Chickweed Lane, Broom-Hilda, Kudzu, Lola, Quigmans, Rubes, Boonies,
  Juilet Jones, Judge Parker, Mark Trail, Henry, Prince Val, New
  Format, Theydoit, Walnut Cove, Tiger, Ripkirby, Hazel, Katzenjammer
  Kids, Mandrake, Pops Place, Popeye, Small Society, Steve Roper,
  Moose Miller, Needhelp, Dick Tracy, Brenda Star, Orphan Annie,
  Gasoline Alley, Soap On A Rope, and World Views
* Added editorial cartoons by: Walt Handelsman,  Bill Day, Chan Lowe,
  Steve Sack, Mike Peters, Dana Summers, Jeff Macnelly, Jack Ohman,
  Don Wright, Dick Locher, Wayne Stayskal
* Changed the default installation location for module files to
  /usr/share/netcomics.
* Added rc file capability.
* Removed Goats for copyright issues.

Major changes in release 0.10:

* Changed from using localtime() throughout the code that downloads
  comics to use gmtime().
* New RLI fields: title, type, author, main, and archives. The field
  called "name" has been depricated. Please read
  Comic_Module-HOWTO.html.
* The created webpages now get extra links included with them: to the
  comic at the website, the comics main page at the website, and the
  website's archives. The body element html template files were
  updated to handle very long headings for each comic.
* The modules have been seperated into their own RPM so that updates
  can be made to them without having to update the the main RPM. The
  name of the RPM is netcomics-data.
* Added the comics: Bliss, The Boondocks, Citizen Dog, The Fusco
  Brothers, Overboard, Tank McNamara, Real Life Adventures, Tradin'
  Paint, Animal Crackers, Bottom Liners, Strange Brew, For Heaven's
  Sake, Heathcliff, & Color Blind.

Major changes in release 0.9:

* The template HTML files now require 8 body_el#.html files where # is
  the number of image files that are handled.  If you have created
  your own HTML template files, be sure to take a look at the ones
  included with this distribution and the document:
  Modify_Webpage_Creation-HOWTO.html
* A new command-line option, -u, has been added.  Using this will
  cause the actual comics to *not* be downloaded.  If you use -W or
  -w, it will make the embedded images link to the comic on its
  website.  If a webpage isn't created, the URLs will be printed on

show_comics.pl
#!/usr/bin/perl -w
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU Library General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.

use strict;
use POSIX;
BEGIN {my $d="/usr/lib/perl5/site_perl"; push(@INC,$d) if ! grep(/^$d$/,@INC);}
use Netcomics::Config;

#Tk loading is delayed until needed so that this script is usable  w/out 
#PerlTk can still use this script

$| = 1;

my $progname = "show_comics";
my @def_time = ("17","00");
my $dont_ask = 0;    
my $cmd = "nice -19 xv";
my $check_display_cmd = "xdpyinfo";
my $rcfile = $ENV{'HOME'} . "/.comics";
my $lockfile = $ENV{'HOME'} . "/.show_comics_lock";
my $check_lock = 1;
my $created_lock = 0;
my @start_time = ();

my $conf = new Netcomics::Config($progname);

##all this CLI needs to be put into a subclass of Netcomics::Config.

my $smushopt = 0;
while (@ARGV)
{
    $_ = shift(@ARGV);

    #set the display command
    if (/^-c$/) {
	if (@ARGV > 0) {
	    $cmd = shift(@ARGV);
	} else {
	    errmsg("Need a command. Use -h for usage.\n");
	    exit 1;
	}
    }
    
    #set the directory
    elsif (/^-d$/) {
	if (@ARGV > 0) {
	    $comics_dir = shift(@ARGV);
	} else {
	    errmsg("Need a directory name. Use -h for usage.\n");
	    exit 1;
	}
    }
    
    #verbosity
    elsif (/^-v/) {
	if ($verbose) {
	    $extra_verbose = 1;
	} else {
	    $verbose = 1;
	}
	$smushopt = 1;
    }
    
    #Specified default time or the leaving time
    elsif (/^-([lt])$/) {
	my $type = $1;
	my $good = 0;
	if (@ARGV > 0) {
	    my $time = shift(@ARGV);
	    if ($time =~ /^(\d\d?):(\d\d?)$/) {
		@def_time = ($1,$2);
		$good = 1;
	    }
	}
	$dont_ask = 1 if $type eq "l";
	unless ($good) {
	    errmsg("Need a time for an argument to -$type. It has the " .
		"form: hh:mm\n. Use -h for usage.\n");
	    exit 1;
	}
    }

    #rc file
    elsif (/^-f$/) {
	if (@ARGV > 0) {
	    $rcfile = shift(@ARGV);
	} else {
	    errmsg("Need a filename for an argument to -f. " .
		"Use -h for usage.\n");
	    exit 1;
	}
    }

    #don't use lock file
    elsif (/^-K/) {
	$check_lock = 0;
	$smushopt = 1;
    }
    
    #specify lock file
    elsif (/^-k$/) {
	if (@ARGV > 0) {
	    $lockfile = shift(@ARGV);
	} else {
	    errmsg("Need a filename for an argument to -k. " .
		"Use -h for usage.\n");
	    exit 1;
	}
    }

    elsif (/^-p(=(.*))?/) {
	#make sure the display is accessible
	my $override_dash = 0;
	my $override_next_opt = 0;
	if (defined($1)) {
	    unshift(@ARGV,$2);
	    infomsg("Pushing $2 onto command line arguments.\n")
		if $extra_verbose;
	    $override_dash = 1 if $2 =~ /^-/;
	} else {
	    $override_next_opt = 1 if /^-p.+/;
	}
	if (@ARGV > 0 && ! $override_next_opt && 
	    ($ARGV[0] !~ /^-/ || $override_dash)) {
	    $check_display_cmd = shift(@ARGV);
	} else {
	    $smushopt = 1;
	}
	infomsg("Running $check_display_cmd to see if display can be " .
		"openned.\n") if $extra_verbose;
	if (system("$check_display_cmd 2>&1 > /dev/null")/256 != 0) {
	    infomsg("$progname: unable to open display.\n");
	    exit 1;
	}
    }
    
    #Specified default time or the leaving time
    elsif (/^-s$/) {
	my $good = 0;
	if (@ARGV > 0) {
	    my $time = shift(@ARGV);
	    if ($time =~ /^(\d\d?):(\d\d?)$/) {
		@start_time = ($1,$2);
		$good = 1;
	    }
	}
	unless ($good) {
	    errmsg("Need a time for an argument to -s. It has the " .
		"form: hh:mm\n. Use -h for usage.\n");
	    exit 1;
	}
    }

    #Usage
    else {
	errmsg("Unknown option: $_.\n");
	usage();
    }
    
    #Allow for smushed-together options
    if ($smushopt) {
	$smushopt = 0;
	if (/^-.(.+)/) {
	    unshift(@ARGV,"-$1");
	    infomsg("Pushing -$1 onto command line arguments.\n") 
		if $extra_verbose;
	}
    }
}

lock_file();

#Create a list of files to display
opendir(DIR,$comics_dir) || 
    die "Could not open the directory to $comics_dir: $!";
my @files = grep(/\.(gif|jpg|jpeg|png)$/,readdir(DIR));
closedir(DIR);
if (-f $rcfile && -r $rcfile) {
    open(FILE,"<$rcfile") || die "Could not open the file $rcfile: $!";
    my @comics = <FILE>;
    close(FILE);
    my @newfiles = ();
    my $comic;
    foreach $comic (grep(/^[^\#]/,@comics)) {
	chomp($comic);
	next if $comic =~ /^$/;
	my @l = grep(/^$comic/,@files);
	push(@newfiles,sort(@l)) if @l > 0;
    }
    @files = @newfiles;
}
if (@files == 0) {
    infomsg("No comics to be displayed. Exiting.\n") if $verbose;
    exit(0);
}
infomsg("Comics to be displayed: @files\n") if $extra_verbose;

#check the time
my $time = time;
my @ltime = localtime($time);
if (@start_time > 0) {
    infomsg("Checking current time against start time.\n") if $extra_verbose;
    if (($ltime[2] == $start_time[0] && $ltime[1] < $start_time[1]) ||
	($ltime[2] < $start_time[0])) {
	infomsg("Exiting because the time ($ltime[2]:$ltime[1]) is before " .
	    "the allowed start time ($start_time[0]:$start_time[1]).\n")
		if $verbose;
	exit(0);
    }
}

#distribute the comics from now until when leaving
my ($hour,$minute) = @def_time;
($hour,$minute) = when_leaving_dialog(@def_time) unless $dont_ask;
infomsg("Time leaving: $hour:$minute\n") if $extra_verbose;
my $stop_time = mktime(0,$minute,$hour,@ltime[3..6]);
my $delaytime = (($stop_time - $time)/@files); 
$delaytime =~ s/(\d+).?\d*/$1/;  #make it a whole number

my $num = @files;
infomsg("Displaying $num comics from now until $hour:$minute, one every " .
    "$delaytime seconds\n") if $verbose;

#display the comics!
if (@files > 0) {
    foreach (@files) {
	my $syscmd = "$cmd $comics_dir/$_";
	infomsg("Running '$syscmd'\n") if $verbose;
	if ($delaytime > 0) {
	    system("$syscmd &");
	    infomsg("Sleeping for $delaytime seconds.\n") if $extra_verbose;
	    sleep $delaytime; 
	} else {
	    system("$syscmd");
	}
    }
}

sub when_leaving_dialog {
    my ($hour,$minute) = ("","");
    ($hour,$minute) = @_ if @_ == 2;

    eval {
	require Tk;
    };
    if ($@) {
	$_ = <<END;
$@
WARNING! Either Tk is not installed or it is not installed correctly.  Use the
-l option next time.  Not querying for the end of the time interval.
END
	errmsg("$_\nDefaulting the time you're leaving to $hour:$minute.\n");
	return ($hour,$minute);
    }
    require Tk::Entry;
    require Tk::Label;

    my $top = MainWindow->new();
    $top->title ("Comics Display Distribution");
    my $label = $top->Label(-text => "When are you leaving?");
    my $entry_hour = $top->Entry(-width => 2);
    my $colon = $top->Label(-text => ":");
    my $entry_minute = $top->Entry(-width => 2);

    my $ok_cb = sub {
	$hour = $entry_hour->get();
	$minute = $entry_minute->get();
	if ($hour =~ /^\d\d?$/ && $hour < 24) {
	    if ($minute =~ /^\d\d?$/ && $minute < 60) {
		$top->destroy();
	    } else {
		errmsg("\nRange Error: Minutes must be from 0 to 59!\n")
		    if $verbose;
		$entry_minute->delete('0.0','end');
	    }
	} else {
	    errmsg("\nRange Error: Hourss must be from 0 to 59!\n")
		if $verbose;
	    $entry_hour->delete('0.0','end');
	}
    };

    my $button = $top->Button(-text => 'OK',-command => [$ok_cb]);
    
    $entry_hour->bind('<Return>' => [$ok_cb]);
    $entry_minute->bind('<Return>' => [$ok_cb]);
    
    $entry_hour->insert('end',$hour);
    $entry_minute->insert('end',$minute);
    
    $label->pack(-side => 'left');
    $entry_hour->pack(-side => 'left');
    $colon->pack(-side => 'left');
    $entry_minute->pack(-side => 'left');
    $button->pack(-side => 'bottom');

    my $x = int(($top->screenwidth - $top->reqwidth)/2 - $top->vrootx);
    my $y = int(($top->screenheight - $top->reqheight)/2 - $top->vrooty);
    $top->geometry("+$x+$y");

    Tk::MainLoop();
    
    $hour = "0$hour" if $hour =~ /^\d$/;
    $minute = "0$minute" if $minute =~ /^\d$/;
    return ($hour,$minute);
}

sub lock_file {
    return unless $check_lock;
    if (@_ == 0) {
	infomsg("Checking for lockfile.\n") if $extra_verbose;
	if (-e $lockfile) {
	    infomsg("Lockfile exists: $lockfile.\nCheck to make sure " .
		    "another show_comics isn't running or \n" .
		    "use -K to ignore the lockfile.\n");
	    exit(0);
	}
	if (open(FILE,">$lockfile")) {
	    print FILE "$$\n";
	    close(FILE);
	    $created_lock = 1;
	    $SIG{'INT'} = \&lock_file;
	    $SIG{'QUIT'} = \&lock_file;
	    $SIG{'KILL'} = \&lock_file;
	    $SIG{'TERM'} = \&lock_file;
	} else {
	    errmsg("Error: could not create lockfile $lockfile.\n" .
		"Either specify a different lockfile with -k or use -K.\n");
	    exit(1);
	}
    } elsif ($created_lock) { #don't remove a lock on END if didn't create it
	$_ = shift;
	infomsg("Received $_. Removing lockfile.\n") if $extra_verbose;
	#any other signal, we delete it & exit
	if (-f $lockfile) {
	    if (unlink($lockfile)) {
		$check_lock = 0;
	    } else {
		errmsg("Could not delete $lockfile.\n") if $verbose;
	    }
	} else {
	    errmsg("Warning: no lockfile ($lockfile).\n") if $verbose;
	}
	exit(0);
    }
}
sub END {
    lock_file("END");
}

sub infomsg {
    print "show_comics: " . $_[0];
}

sub errmsg {
    print STDERR "show_comics: " . $_[0];
}

sub usage {
    print <<END;
$progname:  Show comics downloaded with netcomics, distributed over time
(c)1999 Ben Hochstedler <hochstrb\@cs.rose-hulman.edu>
usage: show_comics [-h] [-c cmd] [-d dir] [-f rcfile] [-l|-t hh:mm] [-p [cmd]]
                   [-k lockfile|-K] [-s hh:mm]
  -c: specify the command used to display each comic strip (def: "nice -19 xv")
  -d: specify the directory where the comics reside (def: /var/spool/netcomics)
  -f: specify the file containing the list of comics to display (def: ~/.comics)
  -h: display usage
  -k: specify the name of the lock file (def: ~/.show_comics_lock)
  -K: don't use a lock file
  -l: specify the time you're going to leave instead of poping up a dialog box
  -p: check to make sure the display is available and if not, quit. optional
      argument is the name of the program to use.  a non-zero return code
      from this program indicates a failure to open the display.
  -s: don't run if the current time is before the specified time.
  -t: specify the default time to be placed in the dialog box
  -v: be verbose

Be sure to set the DISPLAY variable.
END
    exit 1;
}

